[html-aam PR 359] Change figure & figcaption accName computation

pkra opened the pull request:
Moved from https://github.com/w3c/html-aam/pull/359/

closes: https://github.com/w3c/html-aam/issues/564
closes: https://github.com/w3c/html-aam/issues/325

Edit: wpt - https://github.com/web-platform-tests/wpt/pull/49647

# Test, Documentation and Implementation tracking
Once this PR has been reviewed and has consensus from the working group, tests should be written and issues should be opened on browsers. Add N/A and check when not applicable.

* [ ] WPT tests: https://github.com/web-platform-tests/wpt/pull/49647
* [ ] Browser implementations (link to issue or commit):
   * [x] WebKit: https://bugs.webkit.org/show_bug.cgi?id=292490
   * [ ] Gecko: https://bugzilla.mozilla.org/show_bug.cgi?id=2047263
   * [x] Blink: - https://issues.chromium.org/issues/434213625 (just for naming image from figcaption)
* [ ] Does this need AT implementations? No.
* [ ] Related APG Issue/PR:
* [ ] MDN Issue/PR: Needs to be filed as MDN currently states figcaption is accname for figure: https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Elements/figure.

scottaohara committed:
Change figure & figcaption accName computation

resolves #325

This PR revises the mappings for `figure` and `figcaption` to remove the "labelled by" relations, changing them to "details for".
The accessible name computation for `figure` removes `figcaption`, a note is added to further reference this change.

TODO: 
- [ ] need to check on updates to UIA mapping, if any.
- [ ] do we want `figcaption` to map to `caption`? 
- [ ] regardless of above, [`caption`](https://w3c.github.io/aria/#caption) will need to be updated in regards to its referencing `figure` and `figcaption` relationships
- [ ] [`aria-details`](https://w3c.github.io/aria/#aria-details) definition should be updated to reference this change

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
resolve review comments

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
indicate figcaption maps to caption

scottaohara committed:
add in AX details for figure's relationship to figcaption

cc @cookiecrook to make sure i got this right.  thank you

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
finish up mapping detials

verified the UIA mappings with help from @benbeaudry 
made some of the wording consistent between mappings
added in a SHOULD to indicate that if a figure is explicitly provided an accessible name from its figcaption, then do not expose the details relationship, as that'd add redundancy.

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
attempt to fix build error

saying it doesn't know what accname is... so, replacing the instances of the links which may be causing the issue?

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
fix the broken accname/desc references

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
add comment regarding aria-details

see https://bugs.chromium.org/p/chromium/issues/detail?id=1426613#c4

aria-details with an empty string could be useful to authors if there was a specific need to _not_ expose a child figcaption - e.g., if the elements need to have remapped roles for whatever reason.

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
update based on review suggestions

cc @cookiecrook

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

pkra committed:
editorial: move html-aam PR359

pkra commented:
Now pointing to main.

rahimabdi commented on html-aam/index.html:
```diff
@@ -2251,38 +2251,45 @@ <h4 id="el-figcaption">`figcaption`</h4>
-              <td></td>
+              <td>
+                <p>
+                  Only the first child instance of a <a href="#el-figcaption">`figcaption`</a> element that is present in the accessibility tree (e.g., not 'hidden') provides details for its parent
```
@jcsteh's [migrated comment](https://github.com/w3c/html-aam/pull/359/files#r1429366035) provided below:

> For figure, we're explicit that aria-labelledby, aria-describedby or aria-details prevent exposure of the details relation. However, we're not similarly explicit here for the reverse relation. Common sense will hopefully prevail, but I don't love relying on common sense for a spec. Is there something I'm missing that would otherwise make this explicit?
> 
> This is a similar situation to https://github.com/w3c/html-aam/pull/481/files#r1424943049. It seems we tend to use ARIA to express roles in a cross-platform manner wherever possible now, but we don't do that for relations, I guess largely because reverse relations and the details-roles attribute complicate things.

rahimabdi commented:
Re-adding reviewers @cookiecrook @aleventhal @jcsteh.

aleventhal commented on html-aam/index.html:
```diff
@@ -2251,38 +2251,45 @@ <h4 id="el-figcaption">`figcaption`</h4>
                 <div class="relations">
                   <span class="type">Relations:</span>
-                  `ATK_RELATION_LABEL_FOR` with parent <a href="#el-figure">`figure`</a> element
+                  If this is the first instance of a child `figcaption` to its parent <a href="#el-figure">`figure`</a> element, then expose `RELATION_DETAILS_FOR` referencing the parent `figure`
```
@scottaohara @jcsteh 
JAWS turned off their details relation setting by default because it made some pages overly verbose. Is it really necessary when the figcaption is most likely just a sibling of the image being described anyway? Should we file a separate issue for this to remove the automatic assignment of the details relation for it? I don't think I understand why we need it.

scottaohara replied:
i'm not opposed to removing the details relationship.  i think you're right here.  i think what's most important here is to make sure it's clear what the figure content is vs the caption.  the divide of where each start/end.  but the details relationship isn't needed for that.

scottaohara replied:
if people are good with removing this, then i can pull all that out of this PR.  it'd definitely make it a lot simpler.

jcsteh replied:
If we remove the details relationship, am I correct that the only thing this PR will do is remove the label association between figure and figcaption?

> i think what's most important here is to make sure it's clear what the figure content is vs the caption. the divide of where each start/end.

How will this PR achieve that? Or am I misunderstanding something here?

scottaohara replied:
@jcsteh right now if we don't do anything else but remove the detail relationship from the PR, then yes, this PR would simply be about removing figcaption from contributing to the accName of the figure.

per my comment on making the distinction between the figure and its caption clear, i'm open to ideas if the existing caption role for the figcaption can't be the delineator that screen readers could use to identify this separation.

jcsteh replied:
Personally, I think the role is a sufficient delineator.

scottaohara committed:
remove details relations for figure/figcaption

scottaohara committed:
chore: prettier

scottaohara commented on html-aam/index.html:
```diff
@@ -2251,38 +2251,31 @@ <h4 id="el-figcaption">`figcaption`</h4>
             <tr>
               <th>Comments</th>
-              <td></td>
+              <td> </td>
```
```suggestion
              <td></td>
```

scottaohara commented on html-aam/index.html:
```diff
@@ -2338,7 +2325,7 @@ <h4 id="el-figure">`figure`</h4>
             <tr>
               <th>Comments</th>
-              <td></td>
+              <td> </td>
```
```suggestion
              <td></td>
```

scottaohara committed:
further removals

scottaohara commented:
Details relationships have been removed from this PR.

So, what this PR does now is 
1. removes the mappings for figure receiving a name from the figure and figcaption elements from their mapping tables
2. removes figcaption from the name calculation steps in that section of the spec, and explicitly states in that section that figcaption does not contribute to name or description unless specified by the author (e.g., via aria)

With these changes, screen readers probably SHOULD, if they don't do so already, make sure that
1. the figure element has its role exposed (announced), whether it has a name or not(?), so that people can be aware of the start/end of a figure
2. the figcaption element SHOULD similarly have its caption role exposed (announced) so it can be made clear where the figure content vs caption content start/end.

Obviously, i'm not saying any screen reader 'must' do what i outlined, but it seemed the simplest enough way to indicate that some sort of changes to make sure people can effectively understand the figure's content vs caption can be made - since now the overly verbose/redundancy of figcaptions as names to figures is removed - and we will no longer be asking a details relationship to be exposed, since that had the unintended effect of trying to solve one verbosity issue with a different once.

aleventhal commented:
Landed the removal of the auto details relation for Chrome/Edge in https://chromium-review.googlesource.com/c/chromium/src/+/5891779

scottaohara commented:
created web platform test for this PR (also added to OP) - https://github.com/web-platform-tests/wpt/pull/49647

cookiecrook commented on html-aam/index.html:
```diff
@@ -2309,24 +2302,18 @@ <h4 id="el-figure">`figure`</h4>
               <th><a href="https://msdn.microsoft.com/en-us/library/ms726297%28v=VS.85%29.aspx">UIA</a></th>
               <td>
                 <div class="role"><span class="type">Role:</span> Use WAI-ARIA mapping</div>
-                <div class="general">
-                  Accessible name derived from `figcaption` according to the <a href="#figure-element-accessible-name-computation">`figure` Element Accessible Name Computation</a>
-                </div>
```
Sorry I missed this... 

> removes figcaption from the name calculation steps in that section of the spec, and explicitly states in that section that figcaption does not contribute to name or description unless specified by the author (e.g., via aria)

What's the reason for this removal? WebKit and Safari implementations agree on using figcaption as the name if none other is provided via ARIA. Seems logical to me to keep this.

scottaohara replied:
here is one of the examples that was referenced as rational for this change [specifically the link to the wicg thread where people are using figcaption to provide transcripts for a video within a figure](https://github.com/w3c/html-aam/pull/359#issuecomment-1017555985)

cookiecrook replied:
yeah, that's a good example

cookiecrook replied:
Scott and I chatted, and are working on a updated that will account for [this HTML text](https://html.spec.whatwg.org/multipage/embedded-content.html#the-img-element):

> 2. If the image is a descendant of a figure element that has a child figcaption element, and, ignoring the figcaption element and its descendants, the figure element has no [flow content](https://html.spec.whatwg.org/multipage/dom.html#flow-content-2) descendants other than [inter-element whitespace](https://html.spec.whatwg.org/multipage/dom.html#inter-element-whitespace) and the img element, then return the contents of the first such figcaption element.

scottaohara replied:
new tests added to account for the quoted html spec text

scottaohara replied:
updated the img element's name calculation to include figcaption per the above text from the HTML spec

cookiecrook reviewed (dismissed):
flagging a section i may have missed earlier

scottaohara committed:
Merge branch 'main' into monorepo_history--html-aam-pr359

scottaohara committed:
Update accessible name calculation algorithm for img to use figcaption

if an img has no other naming mechanism, and it is a descendent of a figure element with a figcaption, but the figure has no other descendants, then the figcaption can be used as the image's accname.

scottaohara committed:
Merge branch 'main' into monorepo_history--html-aam-pr359

scottaohara committed:
chore: prettier

scottaohara committed:
Merge branch 'main' into monorepo_history--html-aam-pr359

rahimabdi commented on html-aam/index.html:
```diff
@@ -16311,13 +16299,19 @@ <h4 id="el-img-name">`img` Element Accessible Name Computation</h4>
-            <li>Otherwise, if there is no `alt` attribute use the `title` attribute.</li>
+            <li>If the <a data-cite="accname-1.2/#dfn-accessible-name">accessible name</a> is still empty and there is no `alt` attribute, use the `title` attribute.</li>
+            <li>
+              If the <a data-cite="accname-1.2/#dfn-accessible-name">accessible name</a> is still empty and there are no `alt` or `title` attributes, and the `img` is a descendent of a `figure`
```
```suggestion
              If the <a data-cite="accname-1.2/#dfn-accessible-name">accessible name</a> is still empty and there are no `alt` or `title` attributes, and the `img` is a descendant of a `figure`
```

scottaohara committed:
Update html-aam/index.html

Co-authored-by: Rahim Abdi <abdi.abdirahim@gmail.com>

rahimabdi commented:
@scottaohara I've added an implementation tracker to this issue, and a link to the WebKit bug. I've submitted a WebKit PR (https://github.com/WebKit/WebKit/pull/45222#issuecomment-2870017030) and came across a question during implementation for `<img>` accname:

> If the accessible name is still empty and there are no `alt` or `title` attributes, and the `img` is a descendant of a `figure` element with a child `figcaption` **but no other flow content or whitespace descendents**, then use the text equivalent computation of the `figcaption` element's subtree.

Should this be "but no other flow content or non-whitespace descendants"? I believe whitespace counts as flow content (and as text nodes), so curious if those should not factor here and should be stripped to ensure the `<figure>` only has an `<img>` and `<figcaption>` as descendants.

scottaohara commented on html-aam/index.html:
```diff
@@ -16311,13 +16299,19 @@ <h4 id="el-img-name">`img` Element Accessible Name Computation</h4>
+            <li>If the <a data-cite="accname-1.2/#dfn-accessible-name">accessible name</a> is still empty and there is no `alt` attribute, use the `title` attribute.</li>
+            <li>
+              If the <a data-cite="accname-1.2/#dfn-accessible-name">accessible name</a> is still empty and there are no `alt` or `title` attributes, and the `img` is a descendant of a `figure`
+              element with a child `figcaption` but no other <a data-cite="html/dom.html#flow-content-2">flow content</a> or
+              <a data-cite="html/dom.html#inter-element-whitespace">whitespace</a> descendents, then use the <a data-cite="accname-1.2/#mapping_additional_nd_te">text equivalent computation</a> of the
```
```suggestion
              element with a child `figcaption` but no other <a data-cite="html/dom.html#flow-content-2">flow content</a> descendants, then use the <a data-cite="accname-1.2/#mapping_additional_nd_te">text equivalent computation</a> of the
```

scottaohara committed:
Update html-aam/index.html

scottaohara committed:
Merge branch 'main' into monorepo_history--html-aam-pr359

scottaohara commented:
bug filed on chromium specifically for the figcaption not providing an image an accessible name - https://issues.chromium.org/issues/434213625
